### PR TITLE
Bug 1344018 - Add ability to execute custom taskcluster actions per job

### PR DESCRIPTION
**Pull request by wlach**
This change is [Reviewable](https://reviewable.io/reviews/mozilla/treeherder/2220)

### CODE REVIEW

File: ui/js/controllers/tcjobactions.js [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+            $scope.input.jsonPayload = JSON.stringify(jsonSchemaDefaults($scope.input.selectedAction.schema), null, 4);
+        };
+
+        $scope.triggerAction = function() {
+            $scope.triggering = true;
+
+            let task = _.clone($scope.input.selectedAction.task);
+            // update the various date properties with evaluations
+            task = thTaskcluster.refreshTimestamps(task);
+            // fixme: it seems like action tasks define these properties as
```
Comment (imbstack) [inline ui/js/controllers/tcjobactions.js]: I think the confusion here is that `refreshTimestamps` was created for `action.yml` rather than `action.json`. I expect we'll want to do something different with the json version since it has the `$fromNow` stuff in it.

Comment (wlach) [inline ui/js/controllers/tcjobactions.js]: Yeah, that's what the commented out code does, I guess I'm just unsure of the advantage of evaluating it dynamically.

Comment (wlach) [inline ui/js/controllers/tcjobactions.js]: I guess I'll just go with the `$fromNow` stuff, since (1) it's there and it would be weird not to use it (2) I assume @jonasfj added it for a reason.

File: ui/js/controllers/tcjobactions.js [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+            let task = _.clone($scope.input.selectedAction.task);
+
+            ['created', 'deadline', 'expires'].map(function(dateProp) {
+                task[dateProp] = tc.fromNow(task[dateProp]['$fromNow']);
+            });
+            task.payload.env['ACTION_INPUT'] = $scope.input.jsonPayload;
+            task.payload.env['ACTION_PARAMETERS'] = null; // fixme: unclear what to set this to
+            task.payload.env['ACTION_TASK'] = null; // fixme: what is this supposed to do?
+            task.payload.env['ACTION_TASK_ID'] = JSON.stringify(job.taskcluster_metadata.task_id);
+            task.payload.env['ACTION_TASK_GROUP_ID'] = JSON.stringify(originalTask.taskGroupId);
```
Comment (jonasfj) [inline ui/js/controllers/tcjobactions.js]: Assuming that `$scope.input.selectedAction` is an entry from the `actions` array of `actions.json`, then this doesn't even remotely implement the specification:
http://gecko.readthedocs.io/en/latest/taskcluster/taskcluster/action-spec.html#action-kind-task

```js
import _ from 'lodash';
import assert from 'assert';
import Ajv from 'ajv';

/////////////// Collect input

// Get static variables, this is the top-level 'variables' property from actions.json
let variables = $scope.actions_json.variables;

// Get the action wish to trigger, this is an entry from the top-level 'actions' array actions.json
let action = $scope.input.selectAction;

// Get input for the action
let input = JSON.parse($scope.input.jsonPayload || 'null'); // null if no input

// taskId of selected task we're triggering this action for, null if no task is selected 
let selectedTaskId = job.taskcluster_metadata.task_id || null;

// task definition of selected task, null if no task is selected
let selectedTask = null; 
if (selectedTaskId) {
  selectedTaskId = await (new tc.Queue()).task(selectedTaskId);
}

// Get the taskGroupId from any task in the result-set we're working on
let taskGroupId;
if (selectedTask) {
  taskGroupId = selectedTask.taskGroupId;
} else {
  taskGroupId = // TODO: Get the taskGroupId from the decision task, from the current result-set
}


/////////////// Sanity checks (UI shouldn't allow these to fail)

// this should never happen you should filter actions to ensure you only
// present supported actions to the user: actions = actions.filter(({kind) => kind === 'task');
assert(action.kind === 'task', 'Unsupported action kind');

// UI should now allows us to trigger an action _for a selected task_ if the
// action doesn't have a context. Ie. it's a global action, not a task specific action. 
assert((action.context && action.context.length > 0) || !selectedTaskId,
       'action has no context, thus accepts no selected task');

// Either we have a selectedTask and selectedTaskId, or neither
assert((selectedTaskId && selectedTask) || (!selectedTaskId && !selectedTask));

// UI should not allow us to trigger an action that isn't relevant for the selected task
assert(_.some(action.context || [], tagset => {
  // if there is some tagset in action.context, such that every key-value pair
  // from the tagset exists in task.tags, then the action is relevant
  return _.every(tagset, (val, key) => selectedTask.tags[key] === val);
}), 'action is not relevant for selectedTask');

// Validate input against schema, if input/schema is given
if (action.schema) {
  assert((new Ajv()).validate(action.schema, input), 'Input does not satisfy schema!');
} else {
  assert(input === null, 'no input is taken by this action, expected null!');
}


/////////////// Instantiate the task template

let render = // import from: https://gist.github.com/jonasfj/f4998c938bf85c5573d1359616bc086c

let actiontaskId = tc.slugid();
let actionTask = render(action.task, _.defaults({}, {
  taskGroupId: taskGroupId,
  taskId: selectedTaskId,
  task: selectedTask,
  input: input,
}, variables); // Important remember to supply static variables


/////////////// Create action task
let queue = tc.Queue({/* credentials from localStorage */});
await queue.createTask(actiontaskId, actionTask);
```

Comment (jonasfj) [inline ui/js/controllers/tcjobactions.js]: Basically, you cannot may any assumptions on the contents of `actions.json` other than those specified in the specification. In this case you some very specific assumptions about the contents and structure if the task template.
We want to decouple not link things tightly...

Anyways, the render stuff should be pretty solid... I'd be happy to help fix any bugs in both template and the `render` function.

File: ui/js/controllers/tcjobactions.js [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+                                alert("Could not find actions.json");
+                                return;
+                            }
+
+                            let actionsUpload = details[0];
+                            return $http.get(actionsUpload.url).then(function(response) {
+                                // only display actions which should be displayed
+                                // in this task's context
+                                $scope.actions = response.data.actions.filter(function(action) {
+                                    return _.any(action.context.map(function(actionContext) {
```
Comment (jonasfj) [inline ui/js/controllers/tcjobactions.js]: `(action.context || [])` as `action.context` is optional, but from the specification lack of presence is equivalent to `[]`.

File: ui/js/controllers/tcjobactions.js [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+                            if (!details.length) {
+                                alert("Could not find actions.json");
+                                return;
+                            }
+
+                            let actionsUpload = details[0];
+                            return $http.get(actionsUpload.url).then(function(response) {
+                                // only display actions which should be displayed
+                                // in this task's context
+                                $scope.actions = response.data.actions.filter(function(action) {
```
Comment (jonasfj) [inline ui/js/controllers/tcjobactions.js]: First you'll want to do:
`actions = actions.filter(({kind}) => kind === 'task')`

We we only support actions with `{kind: 'task'` at this point...

Comment (jonasfj) [inline ui/js/controllers/tcjobactions.js]: This really just prevents breakages when in future versions of the spec new action types are added...

Note: specification says additional kinds may be added and that implementors should ignore kinds they don't understand... As these can be assumed to be from the future.

File: ui/js/controllers/tcjobactions.js [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+                        title: 'artifact uploaded',
+                        value: 'actions.json'}).then(function(details) {
+                            if (!details.length) {
+                                alert("Could not find actions.json");
+                                return;
+                            }
+
+                            let actionsUpload = details[0];
+                            return $http.get(actionsUpload.url).then(function(response) {
+                                // only display actions which should be displayed
```
Comment (jonasfj) [inline ui/js/controllers/tcjobactions.js]: Before you proceed you'll want to check the version number..
```js
if (response.version !== 1) {
  alert("// TODO: nice message explaining that your treeherder version is outdated");
}
```

Also you'll probably want to store the entire result.. `$scope.actions_json = response.data`, just so we have everything...
Notably we'll have to use `response.data.variables` later when instantiating the templates. These are just constant values...

File: ui/js/controllers/tcjobactions.js (hunk 2 above)
Comment (jonasfj) [inline ui/js/controllers/tcjobactions.js]: Sorry, I forgot to substitute in `action_json.variables` in both my first version or the hacks I outlined in a while back in:
https://gist.github.com/jonasfj/c2e7cd4e0602073cb5d0e227fc4c3ec2

File: ui/js/controllers/tcjobactions.js [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+        // prevent closing of dialog while we're triggering
+        $scope.$on('modal.closing', function(event) {
+            if ($scope.triggering) {
+                event.preventDefault();
+            }
+        });
+
+        let decisionTaskGUID = ThResultSetStore.getGeckoDecisionTaskGUID(repoName, resultsetId);
+        if (decisionTaskGUID) {
```
Comment (jonasfj) [inline ui/js/controllers/tcjobactions.js]: If this is the slugid encoding I recommend calling it `decisionTaskId`

File: ui/js/controllers/tcjobactions.js [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+            let originalTaskId = job.taskcluster_metadata.task_id;
+            $http.get('https://queue.taskcluster.net/v1/task/' + originalTaskId).then(
+                function(response) {
+                    originalTask = response.data;
```
Comment (jonasfj) [inline ui/js/controllers/tcjobactions.js]: you might want to store the original task in the `$scope` so you don't have to download again when triggering the action.

File: ui/js/controllers/tcjobactions.js (hunk 6 above)
Comment (jonasfj) [inline ui/js/controllers/tcjobactions.js]: Also if it is the first `decisionTaskId` for the given result-set, then it is also the `taskGroupId` of all tasks in the result-set...
Which we'll use later when instantiating the task, so you may wish to store it under that name...

If it is not the first `decisionTaskId`, but just one of the decision tasks, (there can be multiple if retriggered), then you'll need to fetch the task definition from either this task or the selected task (if any), to get the `taskGroupId`.


Basically: the taskId of the first decision is equal to the `taskGroupId`.

Comment (wlach) [inline ui/js/controllers/tcjobactions.js]: Most of this code is defined elsewhere, so we should fix this elsewhere. I can file a bug about it though.

I'll change the name though.

File: ui/js/controllers/tcjobactions.js (hunk 7 above)
Comment (wlach) [inline ui/js/controllers/tcjobactions.js]: Good idea, done.

File: ui/js/controllers/tcjobactions.js (hunk 5 above)
Comment (wlach) [inline ui/js/controllers/tcjobactions.js]: Ok fixed!

File: ui/js/controllers/tcjobactions.js (hunk 4 above)
Comment (wlach) [inline ui/js/controllers/tcjobactions.js]: I see no "kind" property on the actions. I do see a tag saying "kind: action-callback". Do we need to add this to actions.json?

File: ui/js/controllers/tcjobactions.js (hunk 3 above)
Comment (wlach) [inline ui/js/controllers/tcjobactions.js]: Fixed.

File: ui/js/controllers/tcjobactions.js (hunk 2 above)
Comment (wlach) [inline ui/js/controllers/tcjobactions.js]: I've updated the code to use your algorithm. There were a number of errors in the implementation which I corrected (see the file `tcactions.js` for details).

File: ui/js/controllers/tcjobactions.js (hunk 4 above)
Comment (jonasfj) [inline ui/js/controllers/tcjobactions.js]: Hmm, you right I just tested the published `actions.json` against:
http://gecko.readthedocs.io/en/latest/taskcluster/taskcluster/action-spec.html#formal-specification

And it didn't match, we're missing the `kind: 'task'` property.

Ideally, we should be validating against this schema in the decision task. But I haven't seen a small json schema validator library for node. Most of them are pretty big.

Filed bug https://bugzilla.mozilla.org/show_bug.cgi?id=1345660 to fix the in-tree bug. Thanks :)

File: ui/js/controllers/tcjobactions.js [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+        let decisionTaskId = ThResultSetStore.getGeckoDecisionTaskGUID(repoName, resultsetId);
+        if (decisionTaskId) {
+            let originalTaskId = job.taskcluster_metadata.task_id;
+            $http.get('https://queue.taskcluster.net/v1/task/' + originalTaskId).then(
+                function(response) {
+                    originalTask = response.data;
+                    ThJobDetailModel.getJobDetails({
+                        job_guid: decisionTaskId,
+                        title: 'artifact uploaded',
+                        value: 'actions.json'}).then(function(details) {
```
Comment (jonasfj) [inline ui/js/controllers/tcjobactions.js]: For the record if you have the `taskId` of the decision task you can construct the URL for the actions.json artifact as: 
```js
let queue = new taskcluster.Queue();
let url = queue.buildUrl(queue.getlatestArtifact, taskId, 'public/actions.json');
```

File: ui/js/controllers/tcjobactions.js [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+                            let actionsUpload = details[0];
+                            return $http.get(actionsUpload.url).then(function(response) {
+                                if (response.data.version !== 1) {
+                                    alert("Wrong version of actions.json, can't continue");
+                                    return;
+                                }
+                                $scope.staticActionVariables = response.data.variables;
+                                // only display actions which should be displayed
+                                // in this task's context
+                                $scope.actions = response.data.actions.filter(function(action) {
```
Comment (jonasfj) [inline ui/js/controllers/tcjobactions.js]: we still need to filter on `kind`... there is patch for you on: https://bugzilla.mozilla.org/show_bug.cgi?id=1345660
Adding the `kind` property...
